internal/app/server/open_saves.go

```diff
@@ -539,21 +532,27 @@ func (s *openSavesServer) UploadChunk(stream pb.OpenSaves_UploadChunkServer) err
 		return status.Errorf(codes.InvalidArgument, "SessionId is not a valid UUID string: %v", err)
 	}
 
-	// Create a chunk reference based on the metadata.
+	// Create a chunk reference based on the metadata. Do not add to blobref right away to minimize writes
 	chunk := chunkref.New(blobKey, int32(meta.GetNumber()))
-	if err := s.metaDB.InsertChunkRef(ctx, chunk); err != nil {
+	blob, err := s.metaDB.ValidateChunkRefPreconditions(ctx, chunk)
+	if err != nil {
 		return err
 	}
 
 	writer, err := s.blobStore.NewWriter(ctx, chunk.ObjectPath())
 	if err != nil {
 		return err
 	}
+	deleteOnExit := true
 	defer func() {
 		if writer != nil {
 			writer.Close()
-			// This means an abnormal exit, so make sure to mark the blob as Fail.
-			s.chunkRefFail(ctx, chunk)
+		}
+		if deleteOnExit {
+			// Delete the object from GS if there are errors uploading the data or inserting a chunkref
+			if derr := s.blobStore.Delete(ctx, chunk.ObjectPath()); derr != nil {
```

Comment:
I think this part can be simplified by cancelling the context that is passed into `s.blobstore.NewWriter`. That is, create a new cancellable context there and call `cancel()` instead of making another call to `s.blobstore.Delete`.

Reply:
Do you know if calling cancel() on the new context that is passed to the Writer will do anything once the object has been successfully uploaded to GS?

The upload might succeed but the comparison to the hash values might fail, or the chunkref insert might also fail. So we need the object in GS to be deleted even then.

Reply:
If you cancel the context prior to the writer being closed, then GCS won't save the object. [Source](https://gocloud.dev/howto/blob/#:~:text=In%20some%20cases,failure%20is%20expected.).